docs/DESIGN.md

```diff
@@ -0,0 +1,164 @@
+_Please be familiar with the user guide before reading this document._
+
+This document describes
+
+ * the roadmap of things to be developed
+ * the overall architecture of how we develop linear base
+ * the ideas behind implementing key parts of linear base
+ * rules and advice on contributing
+
+## Roadmap
+
+1. Build out variants of the fundamentals from [`base`] **along with a test
+   suite**:
+
+    * Control ✓
+      * Applicative ✓
+      * Monad ✓
+    * Data
+      * Bifunctor ✓
+      * Bits
+      * Bool
+      * Char
+      * Either
+      * Eq
+      * Fixed
+      * Foldable
+      * Functor ✓
+      * Int (done via Num) ✓
+      * List
+      * Maybe
+      * Ord
+      * Monoid ✓
+      * Ratio
+      * Proxy
+      * Semigroup (done via Num) ✓
+      * String
+      * Tuple
+    * Numeric.Natural
+    * Prelude (almost done)
+
+2. Ground the development in a practical real world example. Investigate and
+   pick from one of these top 3 examples:
+    * [ixy-languages]
+    * [resourceT/bracket]
+    * allocation free Haskell
+
+3. Fix current projects and issues
+    * A mutable data structure sub-library
+    * Push and pull arrays
+    * The use of lenses in Data.Control.Optics
+    * Minor bugs listed as issues
+    * Ensure the integrity of the test suite!
+
+4. Concurrently with (3), build out examples and tutorials in `/examples`
+
+
+### Backlog
+
+ * Other branches: fd/monad-lift and lots-of-optics
+ * When multiplicity is supported, anything that can be made polymorphic should
+   be made polymorphic -- this will make things much more sane.
+ * There are a lot of old branches that need to be reviewed.
+
+## Overall Architecture
+
+The provided data, functions and classes in linear base can be divided into
+three parts.
+
+1. Variants of tools in `base` (or highly-used common haskell tools) in which
+   some function arrows are replaced with linear arrows.
+  * E.g., `Data.Functor.Linear`'s `Functor` has
+  `fmap :: (a #-> b) -> f a #-> f b`.
+2. Re-designed APIs for tools in `base`.
+  * E.g., file IO is done with a linear resource monad, `RIO` in
+    `System.IO.Resource`
+3. New tools you naturally find yourself wanting with linear types.
+  * E.g., an `Unrestricted` data type.
+  * E.g., mutable arrays in `Data.Array.Mutable.Linear`.
+
+### General Implementation Strategy
+
+This is the strategy (that we've followed so far) for developing linear base:
+
+Variants are reasonably simple to implement. You just add linear arrows where
+you would need them if you were trying to use that base facility in a linear
+function. If there are multiple use cases, you might need to think more about
+the design -- as was the case with [functors]. Here's an example:
+
+```haskell
+foldr :: (a #-> b #-> b) -> b #-> [a] #-> b
+foldr f z = \case
+  [] -> z
+  x:xs -> f x (foldr f z xs)
+```
+
+Re-designed APIs are re-designed because the original API had problems that can
```

Comment:
It's not about APIs having problems. It's about different trade-off: proving more things in exchange for linear types stuff (in particular the syntactic overhead).